File: a271(colorful_carrot)/Python/colorful_carrot.py
from sys import stdin, stdout
from typing import Dict, List, Tuple


for line in stdin:
    for i in range(int(line)):
        x, y, z, w, n, m = map(int, stdin.readline().split())
        # index is used as the carrot number,
        # maps to the corresponding weight increament and poison
        change: Dict[str, Tuple[int, bool]] = {
            '0': (0, False), '1': (x, False), '2': (y, False),
            '3': (-z, False), '4': (-w, True)
        }
        # how many time the rabbit is poisoned
        poisons: int = 0
        for day in stdin.readline().split():
            m -= n * poisons
            if m <= 0:
                break

            # Not sure why much slower than the start-end section below.
            # Should've be more efficient since only query once.
            # Maybe the construction time is more costly than query.
            # cur_change: Tuple[int, bool] = change[day]
            # m += cur_change[0]
            # if cur_change[1]:
            #     poisons += 1

            # start
            m += change[day][0]
            if change[day][1]:
                poisons += 1
            # end

        stdout.write(
            f'{m}g\n' if m > 0 else 'bye~Rabbit\n'
        )

File: a271(colorful_carrot)/Java/ColorfulCarrot.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.IOException;


public class ColorfulCarrot {
    public static void main(String[] args) throws IOException {
        try (BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
             BufferedWriter out = new BufferedWriter(new OutputStreamWriter(System.out))) {
            String line = "";
            while ((line = in.readLine()) != null) {
                int numOfCase = Integer.parseInt(line);
                while ((numOfCase--) != 0) {
                    /* x, y, z, w, n, m */
                    String[] dataLine = in.readLine().split(" ");
                    /*
                     * index is used as the carrot number,
                     * maps to the corresponding weight increament
                     */
                    int[] changes = {
                        0, Integer.parseInt(dataLine[0]), Integer.parseInt(dataLine[1]),
                        -1 * Integer.parseInt(dataLine[2]), -1 * Integer.parseInt(dataLine[3])
                    };

                    int poisons = 0;
                    int decreaseByPoison = Integer.parseInt(dataLine[4]);
                    int weight = Integer.parseInt(dataLine[5]);

                    char[] carrots = in.readLine().replaceAll("\\s+", "").toCharArray();
                    for (char carrot : carrots) {
                        weight -= decreaseByPoison * poisons;
                        if (weight <= 0) {
                            break;
                        }
                        if (carrot == '4') {
                            ++poisons;
                        }
                        weight += changes[carrot - 48];  /* char to int */
                    }
                    out.write(
                        (weight > 0 ? String.valueOf(weight) + "g" : "bye~Rabbit")
                        + '\n'
                    );
                }
                out.flush();
            }
        }  /* try block */
    }
}
